### PR TITLE
Improve dataset preparation support + multiresolution prep

### LINKED ISSUE #4: Support multi-resolution/frame bucketing in `prepare_dataset.py` and multi-GPU

### PR DESCRIPTION
**Pull request by a-r-r-o-w**
Fixes #4

### CODE REVIEW

File: training/dataset.py [hunk 1]
```diff
@@ -358,10 +359,13 @@ def _find_nearest_resolution(self, height, width):
...
 
 class BucketSampler(Sampler):
-    def __init__(self, data_source: VideoDataset, batch_size: int = 8, shuffle: bool = True) -> None:
+    def __init__(
+        self, data_source: VideoDataset, batch_size: int = 8, shuffle: bool = True, drop_last: bool = False
+    ) -> None:
         self.data_source = data_source
         self.batch_size = batch_size
         self.shuffle = shuffle
+        self.drop_last = drop_last
```
Comment (sayakpaul) [inline training/dataset.py:368]: How significant is this?

File: training/prepare_dataset.py [hunk 2]
```diff
@@ -419,190 +353,265 @@ def main():
...
+                }
+                file.write(json.dumps(data) + "\n")
+
 
 if __name__ == "__main__":
     args = get_args()
 
-    assert args.height % 16 == 0, "CogVideoX requires input video height to be divisible by 16."
-    assert args.width % 16 == 0, "CogVideoX requires input video width to be divisible by 16."
+    for height in args.height_buckets:
```
Comment (sayakpaul) [inline training/prepare_dataset.py:609]: Could do this in the argument parser too.

File: training/prepare_dataset.py [hunk 3]
```diff
@@ -419,190 +353,265 @@ def main():
...
-            save_latents_and_embeddings(
-                encoded_images,
-                encoded_videos,
-                prompt_embeds,
-                video_paths_usable,
-                prompts_usable,
-                pathlib.Path(args.output_dir),
-                args.save_image_latents,
+                images = images.permute(0, 2, 1, 3, 4)  # [B, C, F, H, W]
+                image_noise_sigma = torch.normal(
```
Comment (sayakpaul) [inline training/prepare_dataset.py:476]: Let's fix this with a seed to have more determinism.

File: training/prepare_dataset.py [hunk 4]
```diff
@@ -419,190 +353,265 @@ def main():
...
+                    mean=-3.0, std=0.5, size=(images.size(0),), device=device, dtype=weight_dtype
+                )
+                image_noise_sigma = torch.exp(image_noise_sigma)
+                noisy_images = images + torch.randn_like(images) * image_noise_sigma[:, None, None, None, None]
```
Comment (sayakpaul) [inline training/prepare_dataset.py:480]: Same let's fix this with a seed.

File: training/prepare_dataset.py [hunk 5]
```diff
@@ -107,122 +156,13 @@ def get_args() -> Dict[str, Any]:
...
         help="Data type to use when generating latents and prompt embeddings.",
     )
+    parser.add_argument("--seed", type=int, default=42, help="Seed for reproducibility.")
+    parser.add_argument(
+        "--num_artifact_workers", type=int, default=4, help="Number of worker threads for serializing artifacts."
+    )
     return parser.parse_args()
 
 
-def load_dataset_from_local_path(
```
Comment (sayakpaul) [inline training/prepare_dataset.py:113]: Where is this going?

Comment (a-r-r-o-w) [reply training/prepare_dataset.py:113]: Do not require any more. Savings is done in `serialize_artifacts` that takes care of everything optionally

File: training/dataset.py (hunk 1 above)
Comment (a-r-r-o-w) [reply training/dataset.py:368]: I think good to have in case one does not want to use incomplete buckets that don't yet have `args.train_batch_size` number of entries

File: training/prepare_dataset.py (hunk 5 above)
Comment (sayakpaul) [reply training/prepare_dataset.py:113]: But why mix the loading utilities with serialization? That feels unintuitive no?

File: training/dataset.py (hunk 1 above)
Comment (sayakpaul) [reply training/dataset.py:368]: Then let's make a note about it in a comment.

File: training/prepare_dataset.py (hunk 5 above)
Comment (a-r-r-o-w) [reply training/prepare_dataset.py:113]: We aren't mixing anything, no? Instead of having these methods, we now load the video dataset by re-using our `VideoDatasetWithResizing` and `VideoDatasetWithResizeAndRectangleCrop`. Once the computation of latents and embeddings are done, we serialize with `serialize_artifacts`. Not sure I understand what you mean...

Comment (sayakpaul) [reply training/prepare_dataset.py:113]: My bad. Clear now.

File: training/prepare_dataset.py (hunk 4 above)
Comment (a-r-r-o-w) [reply training/prepare_dataset.py]: completed

File: training/prepare_dataset.py (hunk 2 above)
Comment (a-r-r-o-w) [reply training/prepare_dataset.py]: completed

File: training/dataset.py (hunk 1 above)
Comment (a-r-r-o-w) [reply training/dataset.py:385]: completed

**Comment by a-r-r-o-w** [comment]
@sayakpaul okay to merge? could you approve if yes?

File: training/dataset.py [hunk 6]
```diff
@@ -377,3 +398,15 @@ def __iter__(self):
...
                 self.buckets[(f, h, w)] = []
+
+        if self.drop_last:
+            return
+
+        for fhw, bucket in list(self.buckets.items()):
+            if len(bucket) == 0:
+                continue
+            if self.shuffle:
+                random.shuffle(bucket)
```
Comment (sayakpaul) [inline training/dataset.py:409]: Could also fix the seed here?

File: training/prepare_dataset.py [hunk 7]
```diff
@@ -58,6 +83,53 @@ def get_args() -> Dict[str, Any]:
...
+    parser.add_argument(
+        "--id_token",
+        type=str,
+        default=None,
+        help="Identifier token appended to the start of each prompt if provided.",
+    )
+    parser.add_argument(
+        "--height_buckets",
+        nargs="+",
+        type=check_height,
```
Comment (sayakpaul) [inline training/prepare_dataset.py:95]: Beautoful!

File: training/prepare_dataset.py [hunk 8]
```diff
@@ -419,192 +379,257 @@ def main():
...
+        num_workers=args.dataloader_num_workers,
+        pin_memory=args.pin_memory,
+    )
 
-        vae = AutoencoderKLCogVideoX.from_pretrained(args.model_id, subfolder="vae", torch_dtype=dtype)
-        vae = vae.to(device)
+    # 3. Prepare models
+    device = f"cuda:{rank}"
+    
+    generator = torch.Generator(device).manual_seed(args.seed)
```
Comment (sayakpaul) [inline training/prepare_dataset.py:470]: Maybe better to always initialize the seed on a CPU.

**Review by sayakpaul — APPROVED** [review]

File: training/dataset.py (hunk 6 above)
Comment (a-r-r-o-w) [reply training/dataset.py:409]: set_seed already does this in the training and preparation scripts

File: training/prepare_dataset.py (hunk 8 above)
Comment (a-r-r-o-w) [reply training/prepare_dataset.py:470]: hmm, setting it to cpu is giving some weird device mismatch errors :thinking: i will try debugging soon, but for now, on a given gpu and same seed, seems to be fully reproducible (but yes i know why cpu is better alternative). todo
